internal/ssa2ast/func.go

```diff
@@ -0,0 +1,1140 @@
+		valueNameMap:       make(map[ssa.Value]string),
+	}
+}
+
+func (fc *funcConverter) getVarName(val ssa.Value) string {
+	if name, ok := fc.valueNameMap[val]; ok {
+		return name
+	}
+
+	name := fc.namePrefix + strconv.Itoa(len(fc.valueNameMap))
+	fc.valueNameMap[val] = name
+	return name
+}
+
+func (fc *funcConverter) convertSignatureToFuncDecl(name string, signature *types.Signature) (*ast.FuncDecl, error) {
+	funcTypeDecl, err := fc.tc.Convert(signature)
+	if err != nil {
+		return nil, err
+	}
+	funcDecl := &ast.FuncDecl{Name: ast.NewIdent(name), Type: funcTypeDecl.(*ast.FuncType)}
+	if recv := signature.Recv(); recv != nil {
+		recvTypeExpr, err := fc.tc.Convert(recv.Type())
+		if err != nil {
+			return nil, err
+		}
+		f := &ast.Field{Type: recvTypeExpr}
+		if recvName := recv.Name(); recvName != "" {
+			f.Names = []*ast.Ident{ast.NewIdent(recvName)}
+		}
+		funcDecl.Recv = &ast.FieldList{List: []*ast.Field{f}}
+	}
+	return funcDecl, nil
+}
+
+func (fc *funcConverter) convertSignatureToFuncLit(signature *types.Signature) (*ast.FuncLit, error) {
+	funcTypeDecl, err := fc.tc.Convert(signature)
+	if err != nil {
+		return nil, err
+	}
+	return &ast.FuncLit{Type: funcTypeDecl.(*ast.FuncType)}, nil
+}
+
+type AstBlock struct {
+	Index   int
+	HasRefs bool
+	Body    []ast.Stmt
+	Phi     []ast.Stmt
+	Exit    ast.Stmt
+}
+
+type AstFunc struct {
+	Vars   map[string]types.Type
+	Blocks []*AstBlock
+}
+
+func isVoidType(typ types.Type) bool {
+	tuple, ok := typ.(*types.Tuple)
+	return ok && tuple.Len() == 0
+}
+
+func isStringType(typ types.Type) bool {
+	return types.Identical(typ, types.Typ[types.String]) || types.Identical(typ, types.Typ[types.UntypedString])
+}
+
+func getFieldName(tp types.Type, index int) (string, error) {
+	if pt, ok := tp.(*types.Pointer); ok {
+		tp = pt.Elem()
+	}
+	if named, ok := tp.(*types.Named); ok {
+		tp = named.Underlying()
+	}
+	if stp, ok := tp.(*types.Struct); ok {
+		return stp.Field(index).Name(), nil
+	}
+	return "", fmt.Errorf("field %d not found in %v", index, tp)
+}
+
+func (fc *funcConverter) castCallExpr(typ types.Type, x ssa.Value) (*ast.CallExpr, error) {
+	castExpr, err := fc.tc.Convert(typ)
+	if err != nil {
+		return nil, err
+	}
+	valExpr, err := fc.convertSsaValue(x)
+	if err != nil {
+		return nil, err
+	}
+	return ah.CallExpr(&ast.ParenExpr{X: castExpr}, valExpr), nil
+}
+
+func (fc *funcConverter) getLabelName(blockIdx int) *ast.Ident {
+	return ast.NewIdent(fmt.Sprintf("%sl%d", fc.namePrefix, blockIdx))
+}
+
+func (fc *funcConverter) gotoStmt(blockIdx int) *ast.BranchStmt {
+	return &ast.BranchStmt{
+		Tok:   token.GOTO,
+		Label: fc.getLabelName(blockIdx),
+	}
+}
+
+func (fc *funcConverter) getAnonFunctionName(val *ssa.Function) (*ast.Ident, error) {
+	parent := val.Parent()
+	if parent == nil {
+		return nil, nil
+	}
+	anonFuncIdx := slices.Index(parent.AnonFuncs, val)
+	if anonFuncIdx < 0 {
+		return nil, fmt.Errorf("anon func %s for call not found", val.Name())
```

Comment:
nit: please use `%q` instead of `%s` in error messages, just in case any strings contain whitespace, quotes, non-printable characters, etc.